python/lsst/daf/persistence/butler.py

```diff
@@ -567,6 +567,64 @@ def datasetExists(self, datasetType, dataId={}, **rest):
                       storageName, datasetType, str(dataId))
         return True
 
+    def _locate(self, datasetType, dataId, write):
+        """Get one or more ButlerLocations and/or ButlercComposites.
+
+        Parameters
+        ----------
+        datasetType : string
+            The datasetType that is being searched for. The datasetType may be followed by a dot and
+            a component name (component names are specified in the policy). IE datasetType.componentName
+
+        dataId : dict or DataId class instance
+            The dataId
+
+        write : bool
+            True if this is a search to write an object. False if it is a search to read an object. This
+            affects what type (an object or a container) is returned.
+
+        Returns
+        -------
+        If write is False, will return either a single object or None. If write is True, will return a list
+        (which may be empty)
+        """
+        repos = self._repos.outputs() if write else self._repos.inputs()
+        locations = []
+        for repoData in repos:
+            # enforce dataId & repository tags when reading:
+            if not write and dataId.tag and len(dataId.tag.intersection(repoData.tags)) == 0:
+                continue
+            components = datasetType.split('.')
+            datasetType = components[0]
+            components = components[1:]
+            location = repoData.repo.map(datasetType, dataId, write=write)
+            if location is None:
+                continue
+            location.datasetType = datasetType # todo is there a better way than monkey patching here?
+            if len(components) > 0:
+                if not isinstance(location, ButlerComposite):
+                    raise RuntimeError("todo msg; location for a dotted datasetType must be a composite.")
+                # replace the first component name with the datasetType
+                components[0] = location.componentInfo[components[0]].datasetType
+                # join components back into a dot-delimited string
+                datasetType = '.'.join(components)
+                location = self._locate(datasetType, dataId, write)
```

Comment:
I don't think this recursion does what you want, since only `components[0]` is ever modified.  That also suggests that you may need to do a test of a datasetType like "foo.bar.baz".  Also, what if the component datasetType contains one or more dots?